server/s3_put_bucket_versioning_action.cc

```diff
@@ -0,0 +1,194 @@
+    put_bucket_version_factory = bucket_body_factory;
+  } else {
+    put_bucket_version_factory =
+        std::make_shared<S3PutBucketVersioningBodyFactory>();
+  }
+  setup_steps();
+}
+
+void S3PutBucketVersioningAction::setup_steps() {
+  s3_log(S3_LOG_DEBUG, request_id, "Setting up the action\n");
+  ACTION_TASK_ADD(S3PutBucketVersioningAction::validate_request, this);
+  ACTION_TASK_ADD(
+      S3PutBucketVersioningAction::validate_request_xml_versioning_status,
+      this);
+  ACTION_TASK_ADD(
+      S3PutBucketVersioningAction::save_versioning_status_to_bucket_metadata,
+      this);
+  ACTION_TASK_ADD(S3PutBucketVersioningAction::send_response_to_s3_client,
+                  this);
+}
+
+void S3PutBucketVersioningAction::validate_request() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  if (request->has_all_body_content()) {
+    bucket_content_string = request->get_full_body_content_as_string();
+    validate_request_body(bucket_content_string);
+  } else {
+    // Start streaming, logically pausing action till we get data.
+    request->listen_for_incoming_data(
+        std::bind(&S3PutBucketVersioningAction::consume_incoming_content, this),
+        request->get_data_length() /* we ask for all */
+        );
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutBucketVersioningAction::consume_incoming_content() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "Consume data\n");
+  if (request->is_s3_client_read_error()) {
+    client_read_error();
+  } else if (request->has_all_body_content()) {
+    bucket_content_string = request->get_full_body_content_as_string();
+    validate_request_body(bucket_content_string);
+  } else {
+    // else just wait till entire body arrives. rare.
+    request->resume();
+  }
+}
+
+void S3PutBucketVersioningAction::validate_request_body(
+    const std::string& content) {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  put_bucket_version_body =
+      put_bucket_version_factory->create_put_bucket_versioning_body(content,
+                                                                    request_id);
+  if (put_bucket_version_body->isOK()) {
+    bucket_version_status = put_bucket_version_body->get_versioning_status();
+    next();
+  } else {
+    std::string s3_error =
+        put_bucket_version_body->get_additional_error_information();
+    if ("OperationNotSupported" == s3_error) {
+      set_s3_error("OperationNotSupported");
+    } else {
+      set_s3_error("MalformedXML");
+    }
+    send_response_to_s3_client();
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutBucketVersioningAction::validate_request_xml_versioning_status() {
+  s3_log(S3_LOG_DEBUG, request_id, "%s Entry\n", __func__);
+
+  if (put_bucket_version_body->validate_bucket_xml_versioning_status(
+          bucket_version_status)) {
+    next();
+  } else {
+    set_s3_error("IllegalVersioningConfigurationException");
```

Comment:
This should be updated to use the same error observed with AWS S3.

Reply:
Agree

Reply:
done